gwpy/timeseries/timeseries.py

```diff
@@ -1388,28 +1388,184 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
         >>> ax = plot.gca()
         >>> ax.set_ylabel('Strain Amplitude at 331.3 Hz')
         >>> plot.show()
+
+        See also
+        --------
+        TimeSeries.heterodyne
+            for the underlying heterodyne detection method
+        """
+        # stride through the TimeSeries and heterodyne at a single frequency
+        phase = (2 * numpy.pi * f *
+                 self.dt.decompose().value *
+                 numpy.arange(0, self.size))
+        out = self.heterodyne(phase, stride=stride, singlesided=True)
+        if exp:
+            return out
+        mag = out.abs()
+        phase = type(mag)(numpy.angle(out, deg=deg))
+        phase.__array_finalize__(out)
+        phase.override_unit('deg' if deg else 'rad')
+        return (mag, phase)
+
+    def heterodyne(self, phase, stride=1, singlesided=False):
+        """Compute the average magnitude and phase of this `TimeSeries`
+        once per stride after heterodyning with a given phase series
+
+        Parameters
+        ----------
+        phase : `array_like`
+            an array of phase measurements (radians) with which to heterodyne
+            the signal
+
+        stride : `float`, optional
+            stride (seconds) between calculations, defaults to 1 second
+
+        singlesided : `bool`, optional
+            Boolean switch to return single-sided output (i.e., to multiply by
+            2 so that the signal is distributed across positive frequencies
+            only), default: False
+
+        Returns
+        -------
+        out : `TimeSeries`
+            magnitude and phase trends, represented as
+            `mag * exp(1j*phase)` with `dt=stride`
+
+        Notes
+        -----
+        This is similar to the :meth:`~gwpy.timeseries.TimeSeries.demodulate`
+        method, but is more general in that it accepts a varying phase
+        evolution, rather than a fixed frequency.
+
+        Unlike :meth:`~gwpy.timeseries.TimeSeries.demodulate`, the complex
+        output is double-sided by default, so is not multiplied by 2.
+
+        Examples
+        --------
+        Heterodyning can be useful in examining quasi-monochromatic signals
```

Comment:
@mattpitkin, thanks for putting together this example, however as you pointed out earlier there are a couple of issues:

* The example is rather complicated, and
* The example uses routines from LAL that are beyond the scope of gwpy's documentation

Is it possible to split the difference and think of either a simpler use case that doesn't depend on LAL? For example, maybe try constructing a toy model for fdot/f of the 60 Hz U.S. power grid line, then heterodyne some O2 data with that model? Alternatively, you could just inject a pulsar signal (using `TimeSeries.inject`) into O2 data and then recover it with `TimeSeries.heterodyne`?

Reply:
@alurban I've simplified the example to inject a sinusoid with a frequency that evolves with time. I then extract the signal's amplitude using `heterodyne` and compare it to demodulation at a fixed frequency.